runtime/debugger.c

```diff
@@ -138,10 +138,16 @@ static void open_connection(void)
                      caml_strerror(errno, buf, sizeof(buf)));
   dbg_in = caml_open_descriptor_in(dbg_socket);
   dbg_out = caml_open_descriptor_out(dbg_socket);
-  /* The code in this file does not bracket channel I/O operations with
-     Lock and Unlock, but this is safe because the debugger only works
-     with single-threaded programs.  The program being debugged
-     will abort when it creates a thread. */
+  /* We need to lock the channels because the low-level I/O calls expect
+     to work on locked channels. We keep them locked because they are
+     private to this code and never used concurrently. */
+  Lock (dbg_in);
+  Lock (dbg_out);
+  /* We also need to remove them from the list of all open channels,
+     otherwise the flush-everything-at-exit code will try to relock them. */
```

Comment:
I'm surprised.  `caml_open_descriptor_{in,out}` does not set the `CHANNEL_FLAG_MANAGED_BY_GC` flag, so `caml_ml_out_channels_list` should not include these channels in its result list.  What's going on?

Reply:
I don't see any mention of `CHANNEL_FLAG_MANAGED_BY_GC` in `caml_{link,unlink}_channel`. The list of all open channels is used in `caml_ml_out_channels_list`, which is used to implement `Stdlib.flush_all`, which is itself used in `at_exit` to flush everything when the program exits.

Reply:
Oh I missed that. It's a strange interaction between memory management and `flush_all`.

Reply:
Well, there's the channels created from OCaml, which are subject to finalization and closed automatically when the program exits, and there's the channels created by the runtime system for its own uses (loading data from bytecode executables; communicating with the debugger), which are neither finalized nor auto-closed.  So it's not absurd to use a single flag `MANAGED_BY_GC` to govern both finalization and `flush_all`/`close_all`...